### PR TITLE
[#3] Suggest function

### PR DESCRIPTION
**Pull request by willbasky**
Resolve [#3](https://github.com/kowainik/hintman/issues/3)
Yes, It should to add `removeTrailLines` function, but what else?

### CODE REVIEW

**Comment by willbasky** [comment]
`getTrailLines` works not well yet.

File: src/Hintman/Suggest.hs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+import Data.IntMap.Strict (IntMap, traverseWithKey)
+
+type Lines = IntMap Line
+
+data LineState = Plain | Trailing
+    deriving (Eq, Show)
+
+data Line = Line
+    { lName  :: Text
```
Comment (vrom911) [inline src/Hintman/Suggest.hs]: Hmm, is it FilePath? Why do we need to store it for each line?
I guess it's better to find trailing spaces for each file separately. What do you think?

Comment (willbasky) [reply src/Hintman/Suggest.hs]: > What do you think about it?

I think it is cool suggestion! :)

Comment (willbasky) [reply src/Hintman/Suggest.hs]: > What do you think?

`lName` is lagacy code))

File: src/Hintman/Suggest.hs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+toLines :: Text -> [Line]
+toLines txt = map (\(n,x) -> Line {lineRaw = n, lineBody = x, lineState = lineState' x}) lTxt
+  where
+    lTxt :: [(Int, Text)]
+    lTxt = zip [1..] (lines txt)
+
+    lineState' :: Text -> LineState
+    lineState' x
```
Comment (sphaso) [inline src/Hintman/Suggest.hs]: I feel like the `toLines` function is very powerful and other hints might make use of it. Could it make sense to leave the trailing space logic out of it? same for the `suggest` function below, which seems to be for trailing space only

Comment (willbasky) [reply src/Hintman/Suggest.hs]: > to leave the trailing space logic out of it?

Sorry. Do you mean to make `toLines :: Text -> [Line]` with _Trailing_ lines only or ...?

Comment (sphaso) [reply src/Hintman/Suggest.hs]: I was thinking that `toLines` could be a transformation function only (from Text to Line) and then have a separate function (suggestNoTrailingSpace?) to turn the Line into a Suggestion. This way other hints can use `toLines`. But it's just an idea :)

Comment (willbasky) [reply src/Hintman/Suggest.hs]: > (from Text to Line)

Text has multiply lines, `Line` is about one row. 
But yes, I may code function like `suggest :: FilePath -> Text -> [Suggestion]`
Just a sec..

File: src/Hintman/Suggestion/Core.hs [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+       ) where
+
+
+data LineState = Plain | Trailing
+    deriving (Eq, Show)
+
+data Line = Line
+    { lineRaw   :: Int
+    , lineBody  :: Text
+    , lineState :: LineState
```
Comment (vrom911) [inline src/Hintman/Suggestion/Core.hs]: I guess this is not needed in this data type. `Line` should contain only info about the line itself.

File: src/Hintman/Suggestion/Core.hs [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+       , LineState (..)
+       , Suggestion (..)
+       ) where
+
+
+data LineState = Plain | Trailing
+    deriving (Eq, Show)
+
+data Line = Line
+    { lineRaw   :: Int
```
Comment (vrom911) [inline src/Hintman/Suggestion/Core.hs]:
```suggestion
    { lineRow   :: Int
```

File: src/Hintman/Suggestion/TrailingSpaces.hs [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+       , toLines
+       , suggest
+       , suggestRaw
+       ) where
+
+import Data.Text (stripEnd)
+import Hintman.Suggestion.Core (Line(..), LineState(..), Suggestion(..))
+
+
+toLines :: Text -> [Line]
```
Comment (vrom911) [inline src/Hintman/Suggestion/TrailingSpaces.hs]: I guess @sphaso was proposing to have function `toLines` in `Hintman.Suggestion.Core`, but it should only convert `Text` into `[Line]` without changing lines content.
In this module you can have just have a function `FilePath -> [Line] -> [Suggestion]` which will on one bypath check on trailing spaces and write the suggested line.

File: src/Hintman/Suggestion/TrailingSpaces.hs [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    lineState' x
+        | x == stripEnd x = Plain
+        | otherwise = Trailing
+
+getTrailLines :: [Line] -> [Line]
+getTrailLines = filter (\ Line{lineState} -> lineState == Trailing)
+
+suggest :: FilePath -> [Line] -> [Suggestion]
+suggest path list =
+    map (\Line{..} -> Suggestion {suggestionFile = path, suggestionNew = stripEnd lineBody, suggestionComment = Nothing})
```
Comment (vrom911) [inline src/Hintman/Suggestion/TrailingSpaces.hs]: If my previous comment will be applied you won't need to make `stripEnd` twice 🙂

File: src/Hintman/Suggestion/Core.hs [hunk 7]
```diff
@@ -0,0 +1,22 @@
+module Hintman.Suggestion.Core
+       ( Line (..)
+       , LineState (..)
+       , Suggestion (..)
+       ) where
+
+
+data LineState = Plain | Trailing
```
Comment (vrom911) [inline src/Hintman/Suggestion/Core.hs]: I guess after the refactoring this data type won't be needed.

File: src/Hintman/Suggestion/Core.hs [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+data Line = Line
+    { lineRaw   :: Int
+    , lineBody  :: Text
+    , lineState :: LineState
+    } deriving (Eq, Show)
+
+data Suggestion = Suggestion
+    { suggestionFile    :: FilePath
+    , suggestionNew     :: Text
+    , suggestionComment :: Maybe Text  -- ^ Optimal comment to the suggestion
```
Comment (ChShersh) [inline src/Hintman/Suggestion/Core.hs]:
```suggestion
    , suggestionComment :: Maybe Text  -- ^ Optional comment to the suggestion
```

File: src/Hintman/Suggestion/TrailingSpaces.hs [hunk 9]
```diff
@@ -0,0 +1,34 @@
+{-# LANGUAGE NamedFieldPuns #-}
```
Comment (ChShersh) [inline src/Hintman/Suggestion/TrailingSpaces.hs]: I feel that usage of both `RecordWildCards` and `NamedFieldPuns` is redundant. With our naming scheme we shouldn't have problems with using only `RecordWildCards`

File: src/Hintman/Suggestion/Core.hs [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+    } deriving (Eq, Show)
+
+data Suggest = Suggest
+    { suggestFile    :: FilePath
+    , suggestNew     :: Text
+    , suggestComment :: Maybe Text  -- ^ Optional comment to the suggestion
+    } deriving (Eq, Show)
+
+toLines :: Text -> [Line]
+toLines txt = map (\(n,x) -> Line{lineRow = n, lineBody = x}) numberedLines
```
Comment (vrom911) [inline src/Hintman/Suggestion/Core.hs]: If you would name the vars `lineRow` and `lineBoby` you can use `RecordWildCards` to build the `Line`:
```suggestion
toLines txt = map (\(lineRow, lineBody) -> Line{..}) numberedLines
```

File: src/Hintman/Suggestion/TrailingSpaces.hs [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+       ) where
+
+import Data.Text (stripEnd)
+import Hintman.Suggestion.Core (Line (..), Suggest (..))
+
+
+suggest :: FilePath -> [Line] -> [Suggest]
+suggest path =
+    foldr (\Line{..} acc ->
+        if lineBody == stripEnd lineBody then acc
```
Comment (vrom911) [inline src/Hintman/Suggestion/TrailingSpaces.hs]: You should put `stripEnd lineBody` into var, so you won't do this action twice (here and the next line)

File: src/Hintman/Suggestion/TrailingSpaces.hs [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+       ( suggest
+       ) where
+
+import Data.Text (stripEnd)
+import Hintman.Suggestion.Core (Line (..), Suggest (..))
+
+
+suggest :: FilePath -> [Line] -> [Suggest]
+suggest path =
+    foldr (\Line{..} acc ->
```
Comment (vrom911) [inline src/Hintman/Suggestion/TrailingSpaces.hs]: You can use [`mapMaybe`](https://hackage.haskell.org/package/relude-0.3.0/docs/Relude-Monad-Reexport.html#v:mapMaybe) function instead if you will add `Line -> Maybe Suggestion` function.
```haskell
suggest path = mapMaybe foo 
  where
    foo :: Line -> Maybe Suggestion
    foo = ...
```

File: hintman.cabal [hunk 13]
```diff
@@ -26,13 +26,18 @@ library
...
+                       Hintman.Suggestion.TrailingSpaces
 
   build-depends:       base-noprelude >= 4.11 && < 4.12
                      , relude ^>= 0.3.0
                      , servant ^>= 0.14.1
                      , servant-server ^>= 0.14.1
                      , transformers >= 0.3
                      , warp ^>= 3.2
+                     , containers
+                     , text
```
Comment (ChShersh) [inline hintman.cabal]: Let's also sort dependencies in alphabetic order here

File: hintman.cabal [hunk 14]
```diff
@@ -26,13 +26,18 @@ library
                            Hintman.App
                            Hintman.Server
                        Prelude
```
Comment (ChShersh) [inline hintman.cabal]: Let's have `Prelude` either after all modules or before all modules

**Pull request merged**